### PR TITLE
feat: sync outbox dequeue bundling 

### PR DESCRIPTION
**Pull request by matthiasn**
## Summary by CodeRabbit

* **New Features**
  * Outbox message bundling: consecutive text-only messages can be packed into one sync payload; media still send individually. New feature flag added (disabled by default). Max bundle size tunable (default 50).

* **Sync/Delivery**
  * Bundles delivered via sidecar payload files; UI shows “Outbox bundle” payload kind with child count.

* **Documentation**
  * Added implementation plan with rollout and test guidance.

* **Localization**
  * Added translations for cs/de/en/es/fr/ro.

* **Tests**
  * Extensive unit/integration coverage for bundling, DB/repo, sender, UI wiring and edge cases.

### CODE REVIEW

File: lib/features/sync/matrix/sync_event_processor.dart [hunk 1]
```diff
@@ -1896,6 +1896,16 @@ class SyncEventProcessor {
...
+      case SyncOutboxBundle():
+        // Receiver unpacking lands in Step 6 of the outbox bundling plan.
+        // Until then, an OutboxBundle on the wire is treated as an unknown
+        // payload — the calling site already logs and skips when this
+        // returns null without touching the sequence log.
+        _trace(
+          'outboxBundle.skip receiver-unpack not yet implemented',
+          subDomain: 'processor.apply',
+        );
+        return null;
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/matrix/sync_event_processor.dart:1946]: Done in commit 2e4d5a3 (Step 6). The receiver-side unpack now lives in [`lib/features/sync/matrix/outbox_bundle_unpacker.dart`](https://github.com/matthiasn/lotti/blob/main/lib/features/sync/matrix/outbox_bundle_unpacker.dart): `OutboxBundleUnpacker.prepare` resolves the sidecar and recursively prepares each child via the existing per-type pipeline; `OutboxBundleUnpacker.apply` iterates the prepared children and dispatches each through the same per-type apply path used for individually-delivered messages. The processor case at lines 1907–1916 wires those two methods. Tested in [`test/features/sync/matrix/outbox_bundle_unpacker_test.dart`](https://github.com/matthiasn/lotti/blob/main/test/features/sync/matrix/outbox_bundle_unpacker_test.dart) (9 cases) plus an integration smoke test in `sync_event_processor_test.dart`.

File: lib/features/sync/outbox/outbox_repository.dart [hunk 2]
```diff
@@ -74,6 +112,23 @@ class DatabaseOutboxRepository implements OutboxRepository {
...
+        await _database.updateOutboxItem(
+          OutboxCompanion(
+            id: Value(item.id),
+            status: Value(OutboxStatus.sent.index),
+            updatedAt: Value(now),
+          ),
+        );
+      }
+    });
+  }
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/outbox/outbox_repository.dart:124]: Done. Replaced the per-row loop with a single bulk `UPDATE … WHERE id IN (…)` via a new `SyncDatabase.markOutboxItemsSent`. `markRetryBatch` stays per-row because the retry-cap logic is genuinely per-row (each row's incremented retries can independently flip to error).

File: lib/features/sync/outbox/outbox_service.dart [hunk 3]
```diff
@@ -93,6 +93,7 @@ class OutboxService {
           loggingService: _loggingService,
           maxRetriesOverride: maxRetries,
           domainLogger: _domainLogger,
+          bundleMaxSizeProvider: _resolveBundleMaxSize,
         );
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/outbox/outbox_service.dart:97]: Done. The flag is now cached in an `int _bundleMaxSize` field, populated at construction time via `Future.sync(() => journalDb.getConfigFlag(...))` and refreshed via `journalDb.watchConfigFlag(useOutboxBundlingFlag).listen(...)`. `_resolveBundleMaxSize` is now a synchronous field read, so the dequeue hot path pays no microtask cost. Subscription cancelled in `dispose()`.

File: lib/features/sync/outbox/outbox_service.dart [hunk 4]
```diff
@@ -444,9 +449,31 @@ class OutboxService {
...
+  /// Read freshly per drain so flipping the flag at runtime takes effect on
+  /// the next pass without restarting the service.
+  Future<int> _resolveBundleMaxSize() async {
+    try {
+      final enabled = await _journalDb.getConfigFlag(useOutboxBundlingFlag);
+      return enabled ? SyncTuning.outboxBundleMaxSize : 1;
+    } catch (_) {
+      // A transient flag-read failure must never block the outbox: fall
+      // back to the safe single-row path.
+      return 1;
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/outbox/outbox_service.dart]: Resolved by the caching change in #3168934508 — the silent `catch (_)` is gone. The seed read uses `Future.sync(...).then(..., onError: (e, st) => captureException(... 'OUTBOX/bundleFlag.seed' ...))` and the watch subscription has a matching `onError` under `OUTBOX/bundleFlag.watch`. Both error paths covered by tests in `outbox_service_test.dart`.

File: lib/l10n/app_localizations_en.dart [hunk 5]
```diff
@@ -6033,6 +6040,9 @@ class AppLocalizationsEn extends AppLocalizations {
   @override
   String get syncPayloadJournalEntity => 'Journal entry';
 
+  @override
+  String get syncPayloadOutboxBundle => 'Outbox bundle';
+
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/l10n/app_localizations_en.dart:6076]: Done in `lib/features/sync/ui/view_models/outbox_list_item_view_model.dart`: `outboxBundle: (b) => '${messages.syncPayloadOutboxBundle} (${b.children.length})'`. Two new tests in `outbox_list_item_view_model_test.dart` verify the count for both a 3-child bundle and the defensive empty-bundle case (renders `(0)`).

File: lib/features/sync/matrix/matrix_message_sender.dart [hunk 6]
```diff
@@ -517,6 +532,62 @@ class MatrixMessageSender {
...
+        message.jsonPath ?? relativeOutboxBundlePath(uuid.v1());
+    final fullBundleJson = json.encode(
+      message.copyWith(jsonPath: null).toJson(),
+    );
+
+    try {
+      await _savePayloadToDisk(
+        relativePath: relativePath,
+        jsonPayload: fullBundleJson,
+      );
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/matrix/matrix_message_sender.dart]: Done. Added `MatrixMessageSender._isSafeOutboxBundlePath`: only paths that start with `outboxBundlesSegment` (`/outbox_bundles/`) and contain no `..` or `.` segments are honoured. Anything else (incl. `null`-but-tampered-via-deserialization edge cases) falls back to a freshly minted UUID path and the rejection is logged via `captureEvent` under `MATRIX_SERVICE/sendMatrixMsg.outboxBundle.write`.

File: test/features/sync/matrix/matrix_message_sender_test.dart [hunk 7]
```diff
@@ -2365,4 +2365,190 @@ void main() {
...
+    test(
+      'sendOutboxBundlePayloadForTesting returns null when writing the '
+      'sidecar to disk throws — the failure is logged via captureException '
+      'and surfaced as null so the bundle never goes on the wire without '
+      'its children',
+      () async {
+        // Replace the documents directory with a path under /dev/null so the
+        // recursive create at write-time genuinely fails.
+        final unwritable = Directory('/dev/null/cannot-write-here');
+        final brokenSender = MatrixMessageSender(
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply test/features/sync/matrix/matrix_message_sender_test.dart]: Done. The test now creates a regular file under the temp dir and uses it as the parent directory of the sidecar's path. `Directory(...).create(recursive: true)` then throws a `FileSystemException` because the parent is not a directory — works on macOS, Linux, and Windows.

File: test/features/sync/outbox/outbox_processor_test.dart [hunk 8]
```diff
@@ -709,4 +709,716 @@ void main() {
...
+        expect(result.shouldSchedule, isTrue);
+        expect(result.nextDelay, Duration.zero);
+      },
+    );
+
+    test(
+      'Scenario A reproduction (image at position 100 in a 500-row queue): '
+      'the wire pattern is bundle(50) × 2 → image alone → bundle(50) × 8 — '
+      '500 rows ship in 11 sends instead of 500',
+      () async {
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply test/features/sync/outbox/outbox_processor_test.dart:1114]: Done. Updated to `bundle(50) → bundle(49) → image alone → bundle(50) × 8` to match the assertion.

File: test/features/sync/outbox/outbox_processor_test.dart [hunk 9]
```diff
@@ -709,4 +709,716 @@ void main() {
...
+        stubBatchClaimFromQueue(repo, [textItem(id: 1), textItem(id: 2)]);
+        await proc.processQueue();
+
+        // After success, the next bundle on a different subject must start
+        // fresh: simulate a new failing bundle with a different head.
+        stubBatchClaimFromQueue(repo, [
+          textItem(id: 99, subject: 'host:other'),
+        ]);
+        when(() => repo.markRetry(any())).thenAnswer((_) async {});
+        await proc.processQueue();
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply test/features/sync/outbox/outbox_processor_test.dart:1412]: Done. The third `processQueue` pass now re-stubs `sender.send` to return `false`, so the host:other pass actually fails and the assertion has something real to check. Without this, the running counter rolled the third send into a success and the test was effectively asserting on noise.

File: lib/features/sync/outbox/outbox_service.dart [hunk 10]
```diff
@@ -410,6 +411,10 @@ class OutboxService {
           msg: msg,
           commonFields: commonFields,
         ),
+        SyncOutboxBundle() => throw StateError(
+          'SyncOutboxBundle is built at dequeue time and must never be '
+          'enqueued via OutboxService.enqueueMessage.',
+        ),
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/outbox/outbox_service.dart]: Done. The `SyncOutboxBundle` invariant guard now sits at the top of `enqueueMessage`, before the `try` — a buggy caller now gets a propagated `StateError` instead of a logged-and-swallowed enqueue failure. The dispatch arm in the inner switch is reduced to `throw StateError('unreachable')` so the freezed-sealed switch stays exhaustive. The existing test was rewritten to assert `throwsStateError` (instead of `captureException`).

File: docs/implementation_plans/2026-04-30_outbox_message_bundling.md [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+```
+enqueueMessage(msg) ──► Outbox row (priority, filePath?, payloadSize, subject)
+                                    │
+                                    ▼
+                       OutboxProcessor.processQueue()
+                                    │
+                                    ▼
+                        claim()  ──►  send(msg)  ──►  markSent()
+                        (one row)     (one Matrix event)
+```
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply docs/implementation_plans/2026-04-30_outbox_message_bundling.md]: Done. Both diagram fences now have a ```text``` info string.

File: docs/implementation_plans/2026-04-30_outbox_message_bundling.md [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+  String? originatingHostId,
+}) = SyncOutboxBundle;
+```
+
+Run `make build_runner`.
+
+**Invariants enforced in code**:
+- `children` must be non-empty.
+- No child is itself a `SyncOutboxBundle` (no nested bundles).
+- No child carries a media attachment.
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply docs/implementation_plans/2026-04-30_outbox_message_bundling.md:126]: Done. The plan now describes the file-backed transport shape explicitly: inline bundles carry children directly; file-backed bundles travel with `children: []` plus a `jsonPath` referencing the sidecar attachment that holds the real children.

File: lib/features/sync/matrix/matrix_message_sender.dart [hunk 13]
```diff
@@ -203,6 +203,21 @@ class MatrixMessageSender {
...
+        );
+        if (stripped == null) {
+          _trace(
+            'FAIL outboxBundlePayload children=${outboundMessage.children.length}',
+            subDomain: 'matrix.send.error',
+          );
+          return false;
+        }
+        outboundMessage = stripped;
+      }
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/matrix/matrix_message_sender.dart:231]: Good catch — fixed in `MatrixMessageSender.sendMatrixMessage`. Each child of a `SyncOutboxBundle` is now passed through `_ensureOriginatingHostId` before `_sendOutboxBundlePayload` runs. Bundle-delivered journal entities and entry links now carry the same `originatingHostId` metadata as their non-bundled counterparts, so receiver-side sequence tracking works the same in either delivery shape.

File: lib/features/sync/matrix/outbox_bundle_unpacker.dart [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        prepared.add(childPrepared);
+      } on FileSystemException {
+        rethrow;
+      } catch (error, stackTrace) {
+        loggingService.captureException(
+          error,
+          domain: 'MATRIX_SERVICE',
+          subDomain: 'processor.resolve.outboxBundle.child',
+          stackTrace: stackTrace,
+        );
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/matrix/outbox_bundle_unpacker.dart:104]: Done. Broadened the rethrow from `on FileSystemException` to `on IOException` so `SocketException`, `HttpException`, `TlsException`, and `WebSocketException` also bubble up to the caller for retry instead of being log-and-skipped per child. Doc comment updated.

File: lib/l10n/app_ro.arb [hunk 15]
```diff
@@ -1117,6 +1117,8 @@
   "configFlagUseCloudInferenceDescription": "Utilizați servicii AI bazate pe cloud pentru funcții îmbunătățite. Acest lucru necesită o conexiune la internet.",
   "configFlagUseCompressedJsonAttachments": "Comprimă payload-urile JSON de sincronizare",
   "configFlagUseCompressedJsonAttachmentsDescription": "Comprimă cu gzip atașamentele JSON de sincronizare înainte de încărcare. Reduce consumul de lățime de bandă în rețele lente, cu un mic cost CPU la trimitere și primire.",
+  "configFlagUseOutboxBundling": "Grupează mesajele din outbox",
+  "configFlagUseOutboxBundlingDescription": "Împachetează rândurile consecutive doar cu text din outbox într-un singur mesaj de sincronizare, astfel încât rafalele de modificări mici să fie trimise ca un singur mesaj Matrix. Atașamentele media sunt în continuare trimise individual.",
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/l10n/app_ro.arb]: Done. `outbox` → `căsuța de trimitere` in all three Romanian strings: `configFlagUseOutboxBundling`, `configFlagUseOutboxBundlingDescription`, `syncPayloadOutboxBundle`. Regenerated via `make l10n`.

File: lib/features/sync/matrix/outbox_bundle_unpacker.dart [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+      try {
+        await applyChild(child);
+      } catch (error, stackTrace) {
+        loggingService.captureException(
+          error,
+          domain: 'MATRIX_SERVICE',
+          subDomain: 'processor.apply.outboxBundle.child',
+          stackTrace: stackTrace,
+        );
+      }
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply lib/features/sync/matrix/outbox_bundle_unpacker.dart:139]: Done. `OutboxBundleUnpacker.apply` now mirrors the prepare path: `on IOException` rethrows so the parent pipeline can retry the whole bundle, while non-IO errors on a single child are still log-and-skipped (per-child fault isolation for non-retriable errors). Already-applied earlier children stay applied — the receiver's per-type apply path is idempotent on (id, vectorClock), so a redelivery is safe. Doc comment updated and a new unit test in `outbox_bundle_unpacker_test.dart` covers the rethrow behaviour.

File: test/features/sync/outbox/outbox_service_test.dart [hunk 17]
```diff
@@ -5702,4 +5711,174 @@ void main() {
...
+          messageSender: messageSender,
+          processor: processor,
+          activityGate: createGate(),
+          ownsActivityGate: false,
+        );
+        addTearDown(svc.dispose);
+
+        // Drain microtasks so the seed Future and the initial stream
+        // emission populate the cache.
+        await Future<void>.delayed(Duration.zero);
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply test/features/sync/outbox/outbox_service_test.dart]: Done. All four `await Future<void>.delayed(Duration.zero)` calls in `outbox_service_test.dart` are now `await pumpEventQueue()`, matching the repo's fake-time/event-queue policy in `test/README.md`.

File: test/features/sync/outbox/outbox_service_test.dart [hunk 18]
```diff
@@ -5702,4 +5711,174 @@ void main() {
...
+
+    test(
+      'a seed-time getConfigFlag failure is captured under '
+      'OUTBOX/bundleFlag.seed and the cached value stays at the safe '
+      'single-row default — a transient DB hiccup at startup must never '
+      'leave the outbox stuck',
+      () async {
+        when(
+          () => journalDb.getConfigFlag(useOutboxBundlingFlag),
+        ).thenThrow(StateError('transient db'));
```
[the remark this answers is not in the document]
Comment (matthiasn) [reply test/features/sync/outbox/outbox_service_test.dart]: Done. Switched the seed-failure stub from `thenThrow(StateError(...))` to `thenAnswer((_) => Future<bool>.error(StateError(...)))` so the test exercises the constructor's real `.then(onError:)` codepath the way `JournalDb.getConfigFlag` would deliver an I/O failure in production. Test still passes.
